File: backend/app/db/mongo.py
from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase

from app.core.config import settings

client: AsyncIOMotorClient | None = None
database: AsyncIOMotorDatabase | None = None


async def connect_mongo() -> None:
    global client, database
    client = AsyncIOMotorClient(settings.MONGODB_URI)
    database = client[settings.MONGODB_DB]
    await ensure_indexes(database)


async def close_mongo() -> None:
    if client:
        client.close()


def get_database() -> AsyncIOMotorDatabase:
    if database is None:
        raise RuntimeError("MongoDB has not been initialized")
    return database


async def ensure_indexes(db: AsyncIOMotorDatabase) -> None:
    await db.properties.create_index([("location", "2dsphere")])
    await db.properties.create_index([("dedupe_key", 1)], unique=True)
    await db.properties.create_index([("locality_id", 1), ("rent", 1)])
    await db.localities.create_index([("location", "2dsphere")])
    await db.localities.create_index([("slug", 1)], unique=True)
    await db.negotiated_rents.create_index([("property_id", 1), ("created_at", -1)])
    await db.reviews.create_index([("locality_id", 1), ("source", 1)])
    await db.ai_summaries.create_index([("entity_type", 1), ("entity_id", 1)], unique=True)

File: backend/src/db/mongo.ts
import { MongoClient, Db } from "mongodb";
import { config } from "../core/config.js";
import { logger } from "../core/logger.js";

let client: MongoClient | null = null;
let database: Db | null = null;

export async function connectMongo(): Promise<Db> {
  if (database) {
    return database;
  }

  try {
    client = new MongoClient(config.MONGODB_URI);
    await client.connect();
    database = client.db(config.MONGODB_DB);
    await ensureIndexes(database);
    logger.info({ db: config.MONGODB_DB }, "mongodb_connected");
    return database;
  } catch (error) {
    logger.error({ error }, "mongodb_connection_failed");
    throw error;
  }
}

export async function closeMongo(): Promise<void> {
  if (client) {
    await client.close();
    client = null;
    database = null;
    logger.info("mongodb_closed");
  }
}

export function getDatabase(): Db {
  if (!database) {
    throw new Error("MongoDB has not been initialized. Call connectMongo() first.");
  }
  return database;
}

export async function ensureIndexes(db: Db): Promise<void> {
  try {
    await db.collection("properties").createIndex({ location: "2dsphere" });
    await db.collection("properties").createIndex({ dedupe_key: 1 }, { unique: true });
    await db.collection("properties").createIndex({ locality_id: 1, rent: 1 });

    await db.collection("localities").createIndex({ location: "2dsphere" });
    await db.collection("localities").createIndex({ slug: 1 }, { unique: true });

    await db.collection("negotiated_rents").createIndex({ property_id: 1, created_at: -1 });
    await db.collection("reviews").createIndex({ locality_id: 1, source: 1 });
    await db.collection("ai_summaries").createIndex({ entity_type: 1, entity_id: 1 }, { unique: true });
  } catch (error) {
    logger.warn({ error }, "ensure_indexes_warning");
  }
}
